### DISCUSSION TITLE
Vector/Matrix in a CRDT

### DISCUSSION THREAD

**Discussion by JMLX42** [spacejam/rust-crdt#132] [Q&A]
Hi there!

I am working on a new version of [the open source SmartShape Engine](https://github.com/aerys/minko) written 100% in Rust using [wgpu](https://github.com/gfx-rs/wgpu).

One of the main features I am investigating is making the new engine data driven + collaborative in a decentralized way. For example, if a user changes the color of an object, that object would change color for all the other users.

Here is an example that uses some WASM properties to synchronize the state of the WASM modules between multiple instances: https://twitter.com/kettlecorn/status/1620567336704180225 It's very low level and transparent. But IMHO it does cover the elephant in the room for larger projects: consistency and concurrency.

So I am trying to use CRDTs instead. The goal is to store the entire state of the 3D application in a CRDT and communicate the CRDT operations on the network to synchronize that state along multiple instances.

To do that, I need to store things like Vector and Matrix in such CRDT. I have started working with [the crdts crate](https://crates.io/crates/crdts). To implement the state as a `crdts::Map<uuid::Uuid, crdts::Orswot<String, Value>, Value>`. In order for this to work, `Value` must implement `Ord` and `PartialOrd`.

So here is what I came up with:

```
use bytemuck::checked::cast;
use nalgebra::{
    Matrix2, Matrix2x3, Matrix2x4, Matrix3, Matrix3x2, Matrix3x4, Matrix4, Matrix4x2, Matrix4x3,
    Vector2, Vector3, Vector4,
};

type Float32 = ordered_float::OrderedFloat<f32>;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Value {
    I32(i32),
    I64(i64),
    U32(u32),
    U64(u64),
    F32(Float32),
    Vec2i(Vector2<i32>),
    Vec3i(Vector3<i32>),
    Vec4i(Vector4<i32>),
    Vec2u(Vector2<u32>),
    Vec3u(Vector3<u32>),
    Vec4u(Vector4<u32>),
    Vec2f(Vector2<Float32>),
    Vec3f(Vector3<Float32>),
    Vec4f(Vector4<Float32>),
    Mat2x2f(Matrix2<Float32>),
    Mat2x3f(Matrix2x3<Float32>),
    Mat2x4f(Matrix2x4<Float32>),
    Mat3x2f(Matrix3x2<Float32>),
    Mat3x3f(Matrix3<Float32>),
    Mat3x4f(Matrix3x4<Float32>),
    Mat4x2f(Matrix4x2<Float32>),
    Mat4x3f(Matrix4x3<Float32>),
    Mat4x4f(Matrix4<Float32>),
}

impl Value {
    pub fn norm(&self) -> f32 {
        match self {
            Value::I32(v) => *v as f32,
            Value::I64(v) => *v as f32,
            Value::U32(v) => *v as f32,
            Value::U64(v) => *v as f32,
            Value::F32(v) => cast::<Float32, f32>(*v),
            Value::Vec2i(v) => cast::<Vector2<i32>, Vector2<f32>>(*v).norm(),
            Value::Vec3i(v) => cast::<Vector3<i32>, Vector3<f32>>(*v).norm(),
            Value::Vec4i(v) => cast::<Vector4<i32>, Vector4<f32>>(*v).norm(),
            Value::Vec2u(v) => cast::<Vector2<u32>, Vector2<f32>>(*v).norm(),
            Value::Vec3u(v) => cast::<Vector3<u32>, Vector3<f32>>(*v).norm(),
            Value::Vec4u(v) => cast::<Vector4<u32>, Vector4<f32>>(*v).norm(),
            Value::Vec2f(v) => cast::<Vector2<Float32>, Vector2<f32>>(*v).norm(),
            Value::Vec3f(v) => cast::<Vector3<Float32>, Vector3<f32>>(*v).norm(),
            Value::Vec4f(v) => cast::<Vector4<Float32>, Vector4<f32>>(*v).norm(),
            Value::Mat2x2f(v) => cast::<Matrix2<Float32>, Matrix2<f32>>(*v).norm(),
            Value::Mat2x3f(v) => cast::<Matrix2x3<Float32>, Matrix2x3<f32>>(*v).norm(),
            Value::Mat2x4f(v) => cast::<Matrix2x4<Float32>, Matrix2x4<f32>>(*v).norm(),
            Value::Mat3x2f(v) => cast::<Matrix3x2<Float32>, Matrix3x2<f32>>(*v).norm(),
            Value::Mat3x3f(v) => cast::<Matrix3<Float32>, Matrix3<f32>>(*v).norm(),
            Value::Mat3x4f(v) => cast::<Matrix3x4<Float32>, Matrix3x4<f32>>(*v).norm(),
            Value::Mat4x2f(v) => cast::<Matrix4x2<Float32>, Matrix4x2<f32>>(*v).norm(),
            Value::Mat4x3f(v) => cast::<Matrix4x3<Float32>, Matrix4x3<f32>>(*v).norm(),
            Value::Mat4x4f(v) => cast::<Matrix4<Float32>, Matrix4<f32>>(*v).norm(),
        }
    }
}

impl PartialOrd for Value {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Value {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        if self.norm() == other.norm() {
            Ordering::Equal
        } else if self.norm() > other.norm() {
            Ordering::Greater
        } else {
            Ordering::Less
        }
    }
}
```

I know that makes no mathematical sense to use the norm this way. But I don't actually intend to actually sort/compare vectors and/or matrices in my code to compute anything.

It builds just fine.

But I am just wondering if that "trick" doesn't fall short in terms of the expectations the `crdts::Orswot` and `crdts::Map` implementations have of `Ord` and `PartialOrd`. And how to test it properly/thoroughly to make sure the required CRDT properties are indeed preserved.

**Comment by davidrusu** [comment]
Sounds like a fun project!

First I think you may have the the `Map` usage a little off:

> struct Map<K, V, A>;

`K` is the Key of the map, `V` is the value, and `A` is the Actor type.

You set `K=Uuid`, `V=Orswot<..>` `A=Value`

`A` should be a unique actor identifier for each client in your system. e.g. it could be player's id, instead you are using a `Value` type which doesn't really make sense.

#### re: how to test your Ord impl.

Main thing here is to check for transitivity, that is, if a < b and b < c then a < c. This may seem obvious but it's easy to break this property if you are not careful.

The best way to test for this is to use a property testing framework like `quickcheck`. See an example of that here:

https://github.com/rust-crdt/rust-crdt/blob/90d22b7641785c3be0361a5ca07d045d4ac84e45/src/identifier.rs#L285-L299

#### stepping back a bit

Wrapping all of your datatypes into a generic `Value` type is probably the wrong way to go. You'll lose a lot of the type-safety from this since all your fields will now be the same type (`Value`).

Consider moving your CRDT logic up a level to the higher level structures instead. 

For example:

```rust
type Actor = Uuid;
type TimeStamp = (u64, Actor);

struct Entity {
  // each field has it's own crdt container
  position: LWWReg<Vector2<f32>, TimeStamp>
  velocity: LWWReg<Vector2<f32>, TimeStamp>
  inventory: Orswot<Item, Actor>
  //...
}

impl crdts::CvRDT for Entity {
    // Ship the Entity structs between clients and merge those instead of merging
    fn merge(&mut self, other: Entity) {
        self.position.merge(other.position);
        self.velocity.merge(other.velocity);
        self.inventory.merge(other.inventory);
    }
}
```

I obviously don't know the details of how your engine but the main ideas here is: 
- turn the higher-level concepts in your engine into CRDT's
- allow each field of a struct to have the CRDT type that makes sense for it, you can't do this if you put everything into a `Map<..>`.
- Keep the type safety by avoiding overly general Value types

Hope that helps!

**Reply by JMLX42** [reply]
> First I think you may have the the Map usage a little off:

@davidrusu Yes. `crdts::Map<uuid::Uuid, crdts::Orswot<String, Value>, u8>` is obviously most likely closer to what I need. Thank you!

> Wrapping all of your datatypes into a generic Value type is probably the wrong way to go. You'll lose a lot of the type-safety from this since all your fields will now be the same type (Value).

@davidrusu AFAIK I don't loose type-safety, but I will potentially have `try_*` something methods to read a value that might fail a lot yes.

> Consider moving your CRDT logic up a level to the higher level structures instead.

@davidrusu thank you for the example!

Still, If I wanted to follow this approach and broadcast the CRDT changes over the wire, I would need to 👍 

1. write some code that goes through all the (eventually) nested `struct` of the state of the app
2. update such code every time a new field is added to make sure it's state is shared

Which might be error prone and, if the state is big, requires lots of code/processing time even when there are 0 changes. Am I missing something?

If I understand correctly, it is kinda the counterpoint of the Automerge philosophy which features a very "dynamic" JSON-like state CRDT. But in the case of Automerge, we get the CRDT commands for the whole state directly. Correct?
